File: ruby/perfect-numbers/perfect_numbers.rb
class PerfectNumber
  def self.classify(number)
    new(number).classify
  end

  def classify
    if perfect?
      'perfect'
    elsif deficient?
      'deficient'
    elsif abundant?
      'abundant'
    end
  end

  private

  attr_reader :number, :aliquot_sum

  def initialize(number)
    raise if number.negative?

    @number = number
    @aliquot_sum = factors.sum
  end

  def factors
    (1..(number / 2)).filter { (number % _1).zero? }
  end

  def perfect?
    aliquot_sum == number
  end

  def abundant?
    aliquot_sum > number
  end

  def deficient?
    aliquot_sum < number
  end
end

File: clojure/perfect-numbers/src/perfect_numbers.clj
(ns perfect-numbers)

(defn- divisors [n]
  (filter #(zero? (rem n %)) (range 1 (inc (quot n 2)))))

(defn classify [n]
  (when (neg? n) (throw (IllegalArgumentException. "number must be positive")))
  (let [aliquot-sum (apply + (divisors n))]
    (cond
      (> aliquot-sum n) :abundant
      (= aliquot-sum n) :perfect
      (< aliquot-sum n) :deficient)))
